### PR TITLE
Add typed GRUCell with tests

### PR DESCRIPTION
**Pull request by SurajK7**
This PR implements refined type of gruCell in Torch.Typed.Functional,
 GRUCell in Torch.Typed.NN.Recurrent.Cell.GRU,
and adds tests for LSTMCell and GRUCell in Torch.Typed.NN.Recurrent.Cell

### CODE REVIEW

File: hasktorch/test/Torch/Typed/NN/Recurrent/Cell/GRUSpec.hs [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+       (HList
+          '[Parameter '( 'D.CPU, 0) 'D.Float '[21, 5],
+            Parameter '( 'D.CPU, 0) 'D.Float '[21, 7],
+            Parameter '( 'D.CPU, 0) 'D.Float '[21],
+            Parameter '( 'D.CPU, 0) 'D.Float '[21]])
+
+testGRUCell = do
+  let spec = GRUCellSpec  @5 @7 @'D.Float @'( 'D.CPU, 0)
+  model <- A.sample spec
+  pure . flattenParameters $ model
```
Comment (junjihashimoto) [inline hasktorch/test/Torch/Typed/NN/Recurrent/Cell/GRUSpec.hs]: Thank you for your contribution.
Could you add the test comparing typed spec with value-level's one for gruCell?

Comment (SurajK7) [reply hasktorch/test/Torch/Typed/NN/Recurrent/Cell/GRUSpec.hs]: Yeah sure, will do.

Comment (SurajK7) [reply hasktorch/test/Torch/Typed/NN/Recurrent/Cell/GRUSpec.hs]: Yeah sure, will do.

**Comment by austinvhuang** [comment]
LGTM, but might be good for @tscholak to comment as he's most familiar with the typed APIs.

File: hasktorch/test/Torch/Typed/NN/Recurrent/Cell/GRUSpec.hs (hunk 1 above)
Comment (SurajK7) [reply hasktorch/test/Torch/Typed/NN/Recurrent/Cell/GRUSpec.hs]: I have added the tests for gruCell as well as lstmCell, can you please take a look?
https://github.com/SurajK7/hasktorch/blob/cdabac8b6be512ada61b47057ccb3d1419ae3316/hasktorch/test/Torch/Typed/FunctionalSpec.hs#L642-L690
https://github.com/SurajK7/hasktorch/blob/cdabac8b6be512ada61b47057ccb3d1419ae3316/hasktorch/test/Torch/Typed/FunctionalSpec.hs#L975-L997

Comment (junjihashimoto) [reply hasktorch/test/Torch/Typed/NN/Recurrent/Cell/GRUSpec.hs]: Sorry I'm late.
The test is OK. 
Could you push the test for GRUCell branch?

File: hasktorch/test/Torch/Typed/FunctionalSpec.hs [hunk 2]
```diff
@@ -644,6 +644,56 @@ instance
...
+  ( TensorOptions '[4 * hiddenSize, inputSize] dtype device
+  , TensorOptions '[4 * hiddenSize, hiddenSize] dtype device
+  , TensorOptions '[4 * hiddenSize] dtype device
+  , TensorOptions '[4 * hiddenSize] dtype device
+  , ( TensorOptions '[batchSize, hiddenSize] dtype device
+    , TensorOptions '[batchSize, hiddenSize] dtype device
+    )
+  , TensorOptions '[batchSize, inputSize] dtype device
+  , ( TensorOptions '[batchSize, hiddenSize] dtype device
+    , TensorOptions '[batchSize, hiddenSize] dtype device
```
Comment (junjihashimoto) [inline hasktorch/test/Torch/Typed/FunctionalSpec.hs]: @SurajK7 
I think these duplicated constraint is not needed.
Do you have any reason?

Comment (SurajK7) [reply hasktorch/test/Torch/Typed/FunctionalSpec.hs]: Yeah, you're right, sorry about that. Will fix it.

**Comment by junjihashimoto** [comment]
LGTM!

**Pull request merged**